packages/mux-video/src/ads/google-ima-ads-provider.ts

```diff
@@ -0,0 +1,243 @@
+/* eslint @typescript-eslint/triple-slash-reference: "off" */
+/// <reference types="google_interactive_media_ads_types" preserve="true"/>
+
+export type GoogleImaAdsProviderConfig = {
+  customVideoElement: AdsVideoElement;
+  videoElement: HTMLVideoElement;
+  originalSize: DOMRect;
+  adContainer: HTMLElement;
+};
+
+type VideoBackup = {
+  contentTime: number;
+  wasPlaying: boolean;
+  originalSrc: string;
+};
+
+export type AdsVideoElement = HTMLVideoElement & {
+  muxDataKeepSession: boolean;
+  unload(): void;
+};
+
+export class GoogleImaAdsProvider {
+  #adDisplayContainer: google.ima.AdDisplayContainer | undefined;
+  #adsLoader: google.ima.AdsLoader | undefined;
+  #adsManager: google.ima.AdsManager | undefined;
+  #ad: google.ima.Ad | undefined | null;
+  #adProgressData: google.ima.AdProgressData | undefined | null;
+  #adPaused = false;
+  #videoElement: HTMLVideoElement;
+  #customVideoElement: AdsVideoElement;
+  #viewMode: google.ima.ViewMode;
+  #videoBackup: VideoBackup | null = null;
+  #originalSize: DOMRect;
+  #adContainer: HTMLElement;
+
+  constructor(config: GoogleImaAdsProviderConfig) {
+    this.#customVideoElement = config.customVideoElement;
+    this.#videoElement = config.videoElement;
+    this.#viewMode = google.ima.ViewMode.NORMAL;
+    this.#originalSize = config.originalSize;
+    this.#adContainer = config.adContainer;
+  }
+
+  setupAdsManager() {
+    if (!this.#adDisplayContainer) {
+      this.#adDisplayContainer = new google.ima.AdDisplayContainer(this.#adContainer, this.#videoElement);
+      this.#adsLoader = new google.ima.AdsLoader(this.#adDisplayContainer);
+
+      this.#adsLoader.addEventListener(
+        google.ima.AdsManagerLoadedEvent.Type.ADS_MANAGER_LOADED,
+        (adsManagerLoadedEvent: google.ima.AdsManagerLoadedEvent) => {
```

Comment:
more inline event handlers we should be careful about, especially since this stuff is created/torn down in `connectedCallback`